DocGen4/Output/DocString.lean

```diff
@@ -187,56 +249,60 @@ def autoLink (el : Element) : HtmlM Element := do
--- once <https://github.com/leanprover/lean4/issues/4411> is fixed
-
-private def _root_.Lean.Xml.Attributes.toStringEscaped (as : Attributes) : String :=
-  as.fold (fun s n v => s ++ s!" {n}=\"{Html.escape v}\"") ""
-
-mutual
+      let (newContents, newBackrefs) ← modifyContents contents funName backrefs modifyElement
+      return ⟨ ⟨ name, attrs, newContents ⟩, newBackrefs ⟩
 
-private partial def _root_.Lean.Xml.eToStringEscaped : Element → String
-| Element.Element n a c => s!"<{n}{a.toStringEscaped}>{c.map cToStringEscaped |>.foldl (· ++ ·) ""}</{n}>"
-
-private partial def _root_.Lean.Xml.cToStringEscaped : Content → String
-| Content.Element e => eToStringEscaped e
-| Content.Comment c => s!"<!--{c}-->"
-| Content.Character c => Html.escape c
-
-end
+/-- Find all references in a markdown text. -/
+partial def findAllReferences (refsMap : HashMap String BibItem) (s : String) (i : String.Pos := 0)
+    (ret : HashSet String := .empty) : HashSet String :=
+  let lps := s.posOfAux '[' s.endPos i
+  if lps < s.endPos then
+    let lpe := s.posOfAux ']' s.endPos lps
+    if lpe < s.endPos then
+      let citekey := Substring.toString ⟨s, ⟨lps.1 + 1⟩, lpe⟩
+      match refsMap.find? citekey with
+      | .some _ => findAllReferences refsMap s lpe (ret.insert citekey)
+      | .none => findAllReferences refsMap s lpe ret
+    else
+      ret
+  else
+    ret
 
 /-- Convert docstring to Html. -/
-def docStringToHtml (s : String) : HtmlM (Array Html) := do
-  let rendered := match MD4Lean.renderHtml s with
-    | .some res => res
-    | _ => "" -- TODO: should print some error message
-  match manyDocument rendered.mkIterator with
-  | Parsec.ParseResult.success _ res =>
-    -- TODO: use `toString` instead of `eToStringEscaped`
-    -- once <https://github.com/leanprover/lean4/issues/4411> is fixed
-    res.mapM fun x => do return Html.raw <| eToStringEscaped (← modifyElement x)
-  | _ => return #[Html.raw rendered]
+def docStringToHtml (s : String) (funName : String) (backrefs : Array BackrefItem) :
+    HtmlM (Array Html × Array BackrefItem) := do
+  let refsMarkdown := "\n\n" ++ (String.join <|
+    (findAllReferences (← read).refsMap s).toList.map fun s =>
+      s!"[{s}]: references.html#ref_{s}\n")
+  match MD4Lean.renderHtml (s ++ refsMarkdown) with
+  | .some rendered =>
+    match manyDocument rendered.mkIterator with
+    | Parsec.ParseResult.success _ res =>
+      let (newRes, newBackrefs) ← modifyElements res funName backrefs modifyElement
+      -- TODO: use `toString` instead of `eToStringEscaped`
+      -- once <https://github.com/leanprover/lean4/issues/4411> is fixed
+      return (newRes.map fun x => Html.raw (eToStringEscaped x), newBackrefs)
+    | _ =>
+      return (#[.raw "<span style='color:red;'>Error: failed to postprocess: </span>", .raw rendered], backrefs)
+  | .none =>
+    return (#[.raw "<span style='color:red;'>Error: failed to parse markdown: </span>", .text s], backrefs)
```

Comment:
I think this should log too, so that we can easily see any failures on the doc-gen builds.

Reply:
You are right. But seems that `HtmlM` monad doesn't provide access to `IO` so I can't print things to stderr.

Reply:
Done.